### ISSUE TITLE
CQT and resampling normalization: bug or feature?

### ISSUE THREAD

**Issue by bmcfee** [bmcfee/librosa#307]
Tagging @ejhumphrey @ebattenberg @dpwe 

I've mentioned this in a few threads ( #165 #302 ), but wanted to get a focused discussion on this point ahead of 0.4.2.

The CQT method assumes that downsampling the signal and keeping the filter-bank fixed at one octave is a good approximation of applying the full filter-bank to the original signal.  This assumption only holds if the energy of the downsampled signal `y_hat` is scaled to match that of the original `y` after resampling, otherwise, you get exponentially worse approximations going down the octaves. The resamplers used in librosa (either `scipy.signal.resample` or `scikits.samplerate` sinc-interpolation) do not re-scale the output signal, and hence lead to banding artifacts in the resulting spectrograms.

My question for you lot: should we consider this a bug, and fix it in the next minor revision?  (It's at least a thinko.)  Or leave it as is to preserve backwards compatibility?  (FWIW: I think you should be able to simulate the scaling artifacts easily in post-processing, if we really feel the need to support old code.)

I have a PR out #306 that I was planning for 0.5.0 which includes a fix for this, among other changes.  However, if we decide to tag this issue as a bug, I can easily migrate the fix up to here.

**Comment by ejhumphrey** [comment]
I'd think any kind of "banding" is a bug, as you'd want an identical output representation (within some numerical tolerance) whether it was achieved with a clever octave decomposition or the inefficient full-matrix transform. 

FWIW, I _think_ I have memories of getting my own full and octave-decomp CQT implementations (using sox) to be approximately equal, but that would require some fact checking.

**Comment by bmcfee** [comment]
I did a bit more digging into this, and I think some amount of banding artifacts becomes inevitable at the low octaves, due to compounded sampling error.  I've made a notebook demonstrating this [here](https://gist.github.com/8e0ca487d4e9a9da662d), running on the #308 branch, and using different filter-bank normalizations (1, 2, or None).  You can see in the bottom 3 plots that 2 and None both introduce sharp discontinuities at all octaves, while `norm=1` significantly diminishes that.  ( @ebattenberg already showed this before, but it's a useful sanity check.)

On the current master branch, all normalization modes exhibit the discontinuities at all octave boundaries.

The #308 branch fixes this by adding the following two steps:
- `resample` has a flag to rescale the output signal by `sqrt(ratio)` so that energy in ~= energy out
- `cqt` (and like methods) additionally scale the signal by `sqrt(2)` after each downsampling, which can be interpreted as applying the scaling that would have been used on the filter-bank.

**Comment by ebattenberg** [comment]
I'm all for fixing banding.  You said that notebook was created using #308?  So the "fixed" cqt plot is shown in the notebook with `norm=1`?

**Comment by bmcfee** [comment]
> So the "fixed" cqt plot is shown in the notebook with norm=1?

Yes: in each group of plots (spectrogram and difference peaks) the first is norm=1, second is norm=2, and third is norm=None.

For comparison purposes, here's the [same notebook](https://gist.github.com/bmcfee/b148cfa309025ef187d7) run on current master without the resample-scaling fixes.

**Comment by bmcfee** [comment]
and just for fun, here's [one more run](https://gist.github.com/bmcfee/262557743098935c6ba8) forcing `sinc_best` at all resample points to see how much the downsampling accuracy is hurting us. (Not much.)

**Comment by bmcfee** [comment]
I played around with this a bit more, and it seems like with the scaling fix in place, you can knock out the remaining banding artifacts at the top end by increasing `filter_scale`, that is, trading off temporal resolution for frequency resolution.  This isn't terribly satisfying or surprising, but i think it makes sense.

Does anyone else have any thoughts on this?  If not, please say so and I'll close out this issue.

**Comment by dpwe** [comment]
Sounds right.  So the remaining banding artifacts were coming from aliasing
at the down sampled nyquist; shifting the passbands in each octave further
away from the nyquist reduced them?

  DAn.

**Comment by bmcfee** [comment]
Yeah, I think so.  It's a little tricky to see the way that I was measuring discontinuities in the notebooks above, since they discard sign and just look at absolute difference (in log-power space).

If you keep sign and re-run the code thusly:
```python
plt.figure(figsize=(12, 6))
del1 = np.diff(librosa.logamplitude(C1**2, ref_power=np.max), axis=0)
plt.plot(np.mean(del1, axis=1))
plt.axis('tight')
plt.tight_layout()
```
you get this figure (0.4.1)
![image](https://cloud.githubusercontent.com/assets/1190540/13148635/2b8d072e-d62c-11e5-8365-4e39628e2bc1.png)

and this figure (with the fix)
![image](https://cloud.githubusercontent.com/assets/1190540/13148626/184f18be-d62c-11e5-8c56-f3f71cb693fa.png)

(That's with `norm=1`.)

If you increase `filter_scale` by a factor of 4 (~= 0.18 at 21 bins per semitone), it looks like this (0.4.1):

![image](https://cloud.githubusercontent.com/assets/1190540/13148876/138868de-d62d-11e5-97cc-2bd8e9b6f8f4.png)

and this (scale fix)
![image](https://cloud.githubusercontent.com/assets/1190540/13148857/03cc4514-d62d-11e5-9e3f-6880d2c795b0.png)

that is, the discontinuities are either removed (maybe?) or no more obvious than typical differences between frequencies within the same octave.

**Comment by bmcfee** [comment]
Ok, I think we're done on this one.

To simplify everything, I did a sweep of tests on white noise in [this notebook](https://gist.github.com/0446b05ba339e601fba6).  Note that these tests only include `norm=1`, and cqspectra are plotted horizontally to make it easier to visualize the octave gaps.

The tests use an over-sampled frequency basis (5 bins per semitone) and vary the filter scale.  For scale<1, the top end is under-sampled, which produces in-octave banding artifacts.  (S&K avoid this by using redundant filters with different phase offsets -- this seems like more trouble than it's worth to implement at this point.)

Once the signal has been sufficiently downsampled, these artifacts disappear.  Banding also disappears when the filter_scale increases, and becomes worse when scale decreases (as expected).

To sum up: I think there's not much else to do here for this fix.  If either one of the Erics cares to CR #308 and git it a thumbs-up, I'll merge and finalize the 0.4.2 release.
